pkg/workflow/workflow.go

```diff
@@ -0,0 +1,127 @@
+package workflow
+
+import (
+	"context"
+)
+
+// Config is the configuration for constructing a Workflow instance.
+type Config struct {
+	// ErrorHandler is handler called when a workflow experiences an error. The error may originate
+	// from hook or from a task. The original error is alwasy returned from the workflow's Execute.
+	// Optional. Defaults to a no-op handler.
+	ErrorHandler ErrorHandler
+}
+
+// Workflow defines an abstract workflow that can execute a serialized set of tasks.
+type Workflow struct {
+	Config
+	tasks             []Task
+	preWorkflowHooks  []Task
+	postWorkflowHooks []Task
+	preHooks          map[TaskName][]Task
+	postHooks         map[TaskName][]Task
+}
+
+// New initializes a Workflow instance without any tasks or hooks.
+func New(cfg Config) (*Workflow, error) {
+	if cfg.ErrorHandler == nil {
+		cfg.ErrorHandler = nopErrorHandler
+	}
+
+	wflw := &Workflow{
+		Config:    cfg,
+		preHooks:  make(map[TaskName][]Task),
+		postHooks: make(map[TaskName][]Task),
+	}
+
+	return wflw, nil
+}
+
+// AppendTask appends t to the list of workflow tasks.
+func (w *Workflow) AppendTask(t Task) {
+	w.tasks = append(w.tasks, t)
```

Comment:
should we make sure task IDs are unique?

Reply:
We probably should, yes.